### PR TITLE
Artipie Dockerfile includes support of all Artipie storage types.

### LINKED ISSUE #1244: Latest version throwing com.artipie.asto.factory.StorageNotFoundException: Storage with type 's3' is not found
Hello team, I'm trying to setup artipie with S3 (minio) on local using artipie latest version. I found that S3 storage is not found.
When I tried with older version (example: v0.22), itcan successfully store the artifact to S3.

```
artipie  | [vert.x-eventloop-thread-0] INFO com.artipie.http.MainSlice - Failure: java.util.concurrent.CompletionException: com.google.common.util.concurrent.UncheckedExecutionException: com.artipie.asto.factory.StorageNotFoundException: Storage with type 's3' is not found.
artipie  |      at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:687)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:662)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2168)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:144)
artipie  |      at com.artipie.SliceFromConfig.<init>(SliceFromConfig.java:83)
artipie  |      at com.artipie.http.ArtipieRepositories.lambda$resolve$1(ArtipieRepositories.java:111)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:684)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:662)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2168)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:144)
artipie  |      at com.artipie.http.ArtipieRepositories.resolve(ArtipieRepositories.java:103)
artipie  |      at com.artipie.http.ArtipieRepositories.lambda$slice$0(ArtipieRepositories.java:81)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1187)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2309)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:144)
artipie  |      at com.artipie.http.ArtipieRepositories.slice(ArtipieRepositories.java:77)
artipie  |      at com.artipie.http.SliceByPath.response(SliceByPath.java:85)
artipie  |      at com.artipie.http.DockerRoutingSlice.response(DockerRoutingSlice.java:88)
artipie  |      at com.artipie.http.rt.RtRulePath.response(RtRulePath.java:53)
artipie  |      at com.artipie.http.rt.SliceRoute.lambda$response$0(SliceRoute.java:69)
artipie  |      at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
artipie  |      at java.base/java.util.Spliterators$ArraySpliterator.tryAdvance(Spliterators.java:1002)
artipie  |      at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
artipie  |      at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
artipie  |      at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
artipie  |      at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
artipie  |      at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
artipie  |      at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
artipie  |      at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:647)
artipie  |      at com.artipie.http.rt.SliceRoute.response(SliceRoute.java:72)
artipie  |      at com.artipie.http.Slice$Wrap.response(Slice.java:62)
artipie  |      at com.artipie.http.slice.LoggingSlice.lambda$response$1(LoggingSlice.java:71)
artipie  |      at com.artipie.http.SafeSlice$RsSafe.send(SafeSlice.java:79)
artipie  |      at com.artipie.micrometer.MicrometerSlice$MicrometerResponse.send(MicrometerSlice.java:172)
artipie  |      at com.artipie.vertx.VertxSliceServer.serve(VertxSliceServer.java:190)
artipie  |      at com.artipie.vertx.VertxSliceServer.lambda$proxyHandler$1(VertxSliceServer.java:165)
artipie  |      at io.vertx.lang.rx.DelegatingHandler.handle(DelegatingHandler.java:20)
artipie  |      at io.vertx.core.http.impl.Http1xServerRequestHandler.handle(Http1xServerRequestHandler.java:67)
artipie  |      at io.vertx.core.http.impl.Http1xServerRequestHandler.handle(Http1xServerRequestHandler.java:30)
artipie  |      at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:55)
artipie  |      at io.vertx.core.impl.DuplicatedContext.emit(DuplicatedContext.java:158)
artipie  |      at io.vertx.core.http.impl.Http1xServerConnection.handleMessage(Http1xServerConnection.java:145)
artipie  |      at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:157)
artipie  |      at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
artipie  |      at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
artipie  |      at io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.channelRead(WebSocketServerExtensionHandler.java:99)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
artipie  |      at io.vertx.core.http.impl.Http1xUpgradeToH2CHandler.channelRead(Http1xUpgradeToH2CHandler.java:116)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
artipie  |      at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:336)
artipie  |      at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
artipie  |      at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:444)
artipie  |      at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:280)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
artipie  |      at io.vertx.core.http.impl.Http1xOrH2CHandler.end(Http1xOrH2CHandler.java:61)
artipie  |      at io.vertx.core.http.impl.Http1xOrH2CHandler.channelRead(Http1xOrH2CHandler.java:38)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
artipie  |      at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
artipie  |      at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
artipie  |      at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
artipie  |      at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
artipie  |      at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
artipie  |      at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
artipie  |      at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
artipie  |      at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
artipie  |      at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
artipie  |      at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
artipie  |      at java.base/java.lang.Thread.run(Thread.java:833)
artipie  | Caused by: com.google.common.util.concurrent.UncheckedExecutionException: com.artipie.asto.factory.StorageNotFoundException: Storage with type 's3' is not found.
artipie  |      at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2051)
artipie  |      at com.google.common.cache.LocalCache.get(LocalCache.java:3951)
artipie  |      at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4848)
artipie  |      at com.artipie.settings.cache.CachedStorages.storage(CachedStorages.java:72)
artipie  |      at java.base/java.util.Optional.map(Optional.java:260)
artipie  |      at com.artipie.settings.StorageByAlias.lambda$storage$0(StorageByAlias.java:39)
artipie  |      at java.base/java.util.Optional.map(Optional.java:260)
artipie  |      at com.artipie.settings.StorageByAlias.storage(StorageByAlias.java:38)
artipie  |      at com.artipie.settings.cache.StoragesCache.storage(StoragesCache.java:47)
artipie  |      at com.artipie.settings.repo.RepoConfig.lambda$storageOpt$1(RepoConfig.java:182)
artipie  |      at java.base/java.util.Optional.map(Optional.java:260)
artipie  |      at com.artipie.settings.repo.RepoConfig.storageOpt(RepoConfig.java:177)
artipie  |      at com.artipie.settings.repo.RepoConfig.storage(RepoConfig.java:164)
artipie  |      at com.artipie.SliceFromConfig.build(SliceFromConfig.java:230)
artipie  |      at com.artipie.SliceFromConfig.lambda$new$0(SliceFromConfig.java:84)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:684)
artipie  |      ... 80 more
artipie  | Caused by: com.artipie.asto.factory.StorageNotFoundException: Storage with type 's3' is not found.
artipie  |      at com.artipie.asto.factory.Storages.newStorage(Storages.java:70)
artipie  |      at com.artipie.settings.cache.CachedStorages.lambda$storage$0(CachedStorages.java:87)
artipie  |      at com.google.common.cache.LocalCache$LocalManualCache$1.load(LocalCache.java:4853)
artipie  |      at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3529)
artipie  |      at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2278)
artipie  |      at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2155)
artipie  |      at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2045)
artipie  |      ... 95 more
```

Kindly help,
Regards

### PR DESCRIPTION
**Pull request by andpopov**
Closes #1244

### CODE REVIEW

**Comment by ChGen** [comment]
Maybe we should implement it as maven build profile as a next step, to have cleaner implementation?

File: docker-asto/AstoVersion.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        }
+        FileInputStream fis = new FileInputStream(Paths.get(args[0]).toFile());
+        DocumentBuilderFactory builderFactory = DocumentBuilderFactory.newInstance();
+        DocumentBuilder builder = builderFactory.newDocumentBuilder();
+        Document xmlDocument = builder.parse(fis);
+        XPath xPath = XPathFactory.newInstance().newXPath();
+        String expression = ".//dependency/version[../groupId[text() = 'com.artipie'] and ../artifactId[text() = 'asto-core']]";
+        final String value = (String) xPath.compile(expression).evaluate(xmlDocument, XPathConstants.STRING);
+        System.out.print(value);
+        fis.close();
```
Comment (olenagerasimova) [inline docker-asto/AstoVersion.java:31]: Seems like almost each variable can be final and `try-with-recourses` or `try-finally` can be used to close `pom.xml` properly.

**Review by olenagerasimova — CHANGES_REQUESTED** [review]
@andpopov thanks, please, check my comment.

**Comment by dgarus** [comment]
@ChGen 

> Maybe we should implement it as maven build profile as a next step, to have cleaner implementation?

Can additional dependencies be defined in a profile?

**Comment by ChGen** [comment]
> @ChGen
> 
> > Maybe we should implement it as maven build profile as a next step, to have cleaner implementation?
> 
> Can additional dependencies be defined in a profile?

@dgarus , yes, profile may include dependencies:
```
........
  <profiles>
    <profile>
.........
      <repositories>...</repositories>
      <dependencies>...</dependencies>
      <dependencyManagement>...</dependencyManagement>
    </profile>
  </profiles>
```
https://maven.apache.org/pom.html#Profiles

**Comment by dgarus** [comment]
@andpopov @ChGen @olenagerasimova 

The approach suggested in the PR doesn't look user-friendly.

Let's make two maven profiles:
 1. build with s3
 2. build with all storage implementations which we have.

In the release stage, make two docker images – first like now, second with active `build with s3` profile. There is no sense to using something excluding `fs` and `s3` to store artifacts, so we cover all existed cases.

wdyt?

**Comment by ChGen** [comment]
> @andpopov @ChGen @olenagerasimova
> 
> The approach suggested in the PR doesn't look user-friendly.
> 
> Let's make two maven profiles:
> 
>     1. build with s3
> 
>     2. build with all storage implementations which we have.
> 
> 
> In the release stage, make two docker images – first like now, second with active `build with s3` profile. There is no sense to using something excluding `fs` and `s3` to store artifacts, so we cover all existed cases.
> 
> wdyt?

I think first profile should match current build configuration, stay slim in size, and second profile should contain all extra plugins with all their dependencies included, i.e. produce `artipie/artipie-full` docker image.

**Comment by dgarus** [comment]
> second profile should contain all extra plugins

Why do you think it should contain all extra plugins?

**Comment by ChGen** [comment]
> > second profile should contain all extra plugins
> 
> Why do you think it should contain all extra plugins?
Most people now start from artipie with file storage, so this is enough for basic image. Since we develop those extra plugins, we should provide proper way to try them. I think generating docker image for every plugin combination is overkill, so I'd prefer having full image instead. Another alternative would be providing user with script like in this PR to build the image for himself.

**Comment by dgarus** [comment]
> > > second profile should contain all extra plugins
> > 
> > 
> > Why do you think it should contain all extra plugins?
> > Most people now start from artipie with file storage, so this is enough for basic image. Since we develop those extra plugins, we should provide proper way to try them. I think generating docker image for every plugin combination is overkill, so I'd prefer having full image instead. Another alternative would be providing user with script like in this PR to build the image for himself.


This makes sense. I suggest naming for profile somehow neutral and including to it only s3 plugin.
If there will be needing to include something else, we can easily to extend this profile.

**Comment by andpopov** [comment]
I added all storage types to pom.xml and found the total size of artipie jars is increased +11 mega bytes (from 91 MB up to 102MB), so I seam it is reasonable to define all storage type dependencies in pom.xml without using workaround with extra logic.

**Comment by olenagerasimova** [comment]
@andpopov and s3 works in this case? I'm not sure asto-s3 jar is a fat jar with dependencies...

**Comment by andpopov** [comment]
> 

asto-s3 works

**Comment by olenagerasimova** [comment]
Hm, seems reasonable
@dgarus what do you think?

**Review by olenagerasimova — DISMISSED** [review]

**Review by dgarus — DISMISSED** [review]
Thanks! LGTM

**Review by olenagerasimova — APPROVED** [review]

**Comment by olenagerasimova** [comment]
@andpopov thanks

**Pull request merged**